File: String/2472. Maximum Number of Non-overlapping Palindrome Substrings.cpp
/*
2472. Maximum Number of Non-overlapping Palindrome Substrings
*/

// Manacher TC:O(n) SC:O(n)
class Solution {
    vector<int> p;
    string ss;
public:
    int maxPalindromes(string s, int k) {
        int n = s.size();
        int next = 0;
        int res = 0;
        ss = "#";
        for(auto &ch : s) {
            ss += ch;
            ss += "#";
        }

        p.resize(2*n+1);

        while((next = (find(next,k))) != -1) {
            res += 1;
        }

        return res;
    }
    int find(int l, int k) {
        for(int i = l, r = l ,c = l, len; i < p.size(); i++) {
            len = r > i ? min(p[2*c-i], r - i) : 1;
            while(i + len < p.size() && i - len >= l && ss[i+len] == ss[i-len]) {
                if(++len > k) {
                    return i + k + (ss[i+k] == '#' ? 0 : 1);
                } 
            }
            if(i + len > r) {
                r = i + len;
                c = i;
            }
            p[i] = len;
        }

        return -1;
    }
};


// DP TC:O(n^2) SC:O(n^2)
class Solution {
public:
    int maxPalindromes(string s, int k) {
        int n = s.size();
        // DP
        vector<vector<bool>> check(n, vector<bool>(n));
        for(int i = 0; i < n; i ++) {
            check[i][i] = 1;
        }
        for(int i = 1; i < n; i ++) {
            if(s[i] == s[i-1]) {
                check[i-1][i] = 1;
            }
        }
        for(int len = 3; len <= n; len++) {
            for(int i = 0; i + len - 1 < n; i++) {
                int j = i + len - 1;
                if(s[i] == s[j] && check[i+1][j-1]) {
                    check[i][j] = 1;
                }
            }
        }
        
        vector<int> dp(n);
        dp[k-1] = check[0][k-1];
        for(int i = k; i < n; i++) {
            dp[i] = dp[i-1];
            for(int j = i-k+1; j >= 0; j--) {
                dp[i] = max(dp[i], (j-1 >= 0 ? dp[j-1] : 0) + (check[j][i]));
            }
        }

        return dp[n-1];
    }
};

File: Dynamic_Programming/2472. Maximum Number of Non-overlapping Palindrome Substrings.py
# TC:O(n^2) SC:O(n^2) Manacher可以做到O(n)
class Solution:
    def maxPalindromes(self, s: str, k: int) -> int:
        '''
        選擇不重疊的substring 必須長度至少k, 每個都是回文串 -> 區間DP
        最多可以選幾個substring -> 不重疊就需要去根據選擇的string劃分
        '''
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        for i in range(n):
            dp[i][i] = True
            if i + 1 < n:
                dp[i][i+1] = True if s[i] == s[i+1] else False
        for i in range(n-2, -1, -1):
            for j in range(i + 2, n):
                if s[i] == s[j] and dp[i+1][j-1]:
                    dp[i][j] = True
        f = [0] * (n+1)
        for i in range(n):
            if i < k-1:
                continue
            f[i+1] = f[i]
            for j in range(i - k + 1, -1, -1):
                if dp[j][i]:
                    f[i+1] = max(f[i+1], f[j] + 1)
        return f[n]
        

# 中心回文串判定法 可以做到 one pass
class Solution:
    def maxPalindromes(self, s: str, k: int) -> int:
        n = len(s)
        f = [0] * (n + 1)
        for i in range(2 * n - 1):
            l, r = i // 2, (i + 1) // 2
            f[l + 1] = max(f[l + 1], f[l])
            while l >= 0 and r < n and s[l] == s[r]:
                if r - l + 1 >= k:
                    f[r + 1] = max(f[r + 1], f[l] + 1)
                    break
                l -= 1
                r += 1
        return f[n]

        
